### ISSUE TITLE
Steady state kubelet CPU usage is high due to excessive allocation in prometheus scraping of cadvisor

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#104459]
The `/metrics/cadvisor` endpoint uses a custom prometheus collector to generate the many thousands of series necessary to get pod/container level metrics from the kubelet for cpu/memory/container info.  Unfortunately, the pattern here generates a very large number of transient allocations on each scrape.

Over about 10-15 scrapes we generate about 7 million objects which corresponds to about 20% of memory allocations and about 20% (approximately) of CPU during steady state kubelet use on a lightly loaded node (20 pods).

```
   5657009 17.01% 17.01%    5657009 17.01%  k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.makeLabelPairs
   1900758  5.72% 22.73%    2381366  7.16%  k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.NewDesc
```

A normal prometheus metric vector caches the creation of dto.LabelPairs because the default golang protobuf implementation is very pointer happy (`[]*struct{ key *string, value *string}`) which causes a lot of allocations on the heap that are unnecessary and also because most of the time the set of label pairs is consistent from scrape to scrape.

However, because of how cadvisor/metrics/prometheus.go creates those, label pairs aren't cached or reused (semi-intentionally) so every scrape does the maximum number of possible allocations, when in practice 99% of those allocations can be avoided.

The fix is to amortize allocations by reusing desc across collect intervals, but that requires us to track and flush unused descriptors after the containers are out of scope.  That would in theory cut about 20% of kubelet cpu use across most of the fleet that is scraping `metrics/cadvisor` (which is pretty high since these are the basic metrics most distros want for admins).

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by derekwaynecarr** [comment]
@mrunal @ehashman @dchen1107 i will add this to our tracking doc for upcoming release.  thanks for the analysis @smarterclayton

**Comment by mrunalp** [comment]
Some other ideas:
1. There are a bunch of constant metrics which could just be cached and returned from cadvisor. 
2. Look into a LinkedHashMap (or similar data structure) to keep the data in memory that allows O(1) deletes and updates on the container data and avoid some churn.

**Comment by mrunalp** [comment]
For a first iteration, even looping over a hashmap should do.

**Comment by ehashman** [comment]
Scratch notes from reproducing on an OpenShift cluster...

```bash
oc proxy

# scrape some metrics
while true; do curl http://127.0.0.1:8001/api/v1/nodes/ip-10-0-159-99.us-west-2.compute.internal/proxy/metrics >& /dev/null; sleep 5; done

go tool pprof -http localhost:8889 http://127.0.0.1:8001/api/v1/nodes/ip-10-0-159-99.us-west-2.compute.internal/proxy/debug/pprof/heap
```

Concur that MakeLabelPairs/NewDesc are by far the bulk of kubelet allocations:

<table>
<thead>
        <tr>
          <th id="flathdr1">Flat</th>
          <th id="flathdr2">Flat%</th>
          <th>Sum%</th>
          <th id="cumhdr1">Cum</th>
          <th id="cumhdr2">Cum%</th>
          <th id="namehdr">Name</th>
          <th>Inlined?</th>
        </tr>
      </thead>
<tr id="node0" class=""><td>40152794</td><td>17.90%</td><td>17.90%</td><td>40152794</td><td>17.90%</td><td>k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.MakeLabelPairs</td><td></td></tr>
<tr id="node1" class=""><td>16741300</td><td>7.46%</td><td>25.37%</td><td>20542467</td><td>9.16%</td><td>k8s.io/kubernetes/vendor/github.com/prometheus/client_golang/prometheus.NewDesc</td><td></td></tr>
</table>

Graph view:

![Screenshot at 2021-09-22 16-43-57](https://user-images.githubusercontent.com/4706131/134436299-9c28ecc2-ec54-48d1-b26f-b6fcbbcebe6c.png)

So it seems that we are recreating constant metrics over and over again via [MustNewConstMetric](https://github.com/prometheus/client_golang/blob/35ee299e50b8b3a40af7bdc9d9703a22dff0ecb6/prometheus/value.go#L102-L110). Here's the code with all the constant metrics, I believe it looks pretty clear:

https://github.com/google/cadvisor/blob/0549d482f3464c844e217bf77116a236a61b5d66/metrics/prometheus.go#L1899-L1921

As Mrunal suggested, we can probably optimize by not creating new constant metrics every time this is called that could be cached. I think this is the lowest hanging fruit in terms of allocations.

**Comment by ehashman** [comment]
I also discussed with @dashpole and @dgrisonnet today during the SIG Instrumentation Triage meeting. There are a few possible options here:

- We could stop using ephemeral metrics, but then we'd need to implement garbage collection (i.e. unregister metrics when containers disappear)
- We could implement some sort of caching mechanism where we scrape everything on an interval and always return the result of the cache. However, this will cause consistency issues and may even miss short-lived containers
- [what I am currently looking into] We could use an LRU cache for every metric that's logically safe to cache (e.g. start time, cgroups info, etc.) and see how much savings this provides

We're somewhat limited in what we can do because Prometheus [best practices](https://www.robustperception.io/setting-a-prometheus-counter) for [custom collectors](https://prometheus.io/docs/instrumenting/writing_exporters/#collectors) is to use constant (ephemeral) metrics. Thus, there's not much of a way around the complexity of `n_metric_samples * m_containers` on every single scrape.

An LRU cache might give us `(n_metric_samples - k_cached_metrics) * m_containers`. Long term/architecturally, we'd get the most savings out of garbage collection, but we are also [moving away from using cadvisor](https://github.com/kubernetes/enhancements/issues/2371), so I think we should start with the least invasive solution and see where that gets us.

**Comment by ehashman** [comment]
Staring at the labels that look constant: I'm not actually sure if we can cache these wholesale because they could change for a given container, for example "container start time" could change if the container restarts, right?

A lot of the way we do label processing is for safety/consistency; see e.g. https://github.com/google/cadvisor/issues/1704 which explains the `rawLabels` handling in https://github.com/google/cadvisor/pull/1831

The containerLabelFunc we use in k8s is defined here: https://github.com/kubernetes/kubernetes/blob/86d23cf441d2c9f228e9d01bf82deda9eb5c567b/pkg/kubelet/server/server.go#L963-L998

Are there immutability guarantees on the values of container labels for a given ID? For example, I know that Pods can [change container images](https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement), but will a container with a given ID always have the same name/image/pod/namespace/container? Guessing yes, and will try to hack on something under that assumption, but it'd be good to confirm.

**Comment by ehashman** [comment]
Started writing some code to cache all container labels, keyed by container ID. We could save some time just by caching the metric descriptions with labels and avoid calling `NewDesc` repeatedly.

However, the serialization work inside the Prometheus client with `MakeLabelPairs` is doing the bulk of the allocations, so the highest impact way to shave off CPU time would probably be to sit on a trove of pre-computed `Prometheus.Metric`s whose values we can update while leaving the labels unchanged. However, you can't actually do that with a constant metric: https://github.com/prometheus/client_golang/blob/8184d76b3b0bd3b01ed903690431ccb6826bf3e0/prometheus/value.go#L87

I wonder if there's something that still allows us to push ephemeral, "almost-constant" metrics into the collector without having to reconstruct a new constant metric every time.

**Comment by ehashman** [comment]
I think we've hit a fundamental limitation of the Prometheus Go client library. I did a bit of research into how kube-state-metrics (KSM) handled this problem and @dgrisonnet reminded me of the performance optimization that was finished in 2018 (that I should have remembered because I was a scale tester!).

Essentially, KSM dropped the Prometheus client and wrote their own, less expensive [serializer](https://github.com/kubernetes/kube-state-metrics/blob/8dab6f7472c26987ab7f8899a4a2f753fed8e8a8/pkg/metric/metric.go) which is more amenable to caching. More info in the original PR (https://github.com/kubernetes/kube-state-metrics/pull/709) and [design proposal](https://github.com/kubernetes/kube-state-metrics/blob/8dab6f7472c26987ab7f8899a4a2f753fed8e8a8/docs/design/metrics-store-performance-optimization.md). This resulted in significant [CPU and memory savings](https://github.com/kubernetes/kube-state-metrics/issues/498#issuecomment-443872447).

Because this isn't available as an abstract implementation in the Prometheus client, if we wanted to take this approach we'd end up with a NIH rewrite. This is non-trivial, and I am not sure how good the cadvisor test coverage is, so it's risky, especially given that we want to deprecate the usage of cadvisor in k8s long-term.

I think for now, our next best steps will be to test how much savings we can get out of caching `NewDesc` calls and possibly some of the always-constant metrics, since that won't be too invasive. It might be worth taking as a patch while we work on a longer-term solution if that improves performance, even by a small amount. I think the Prometheus team should consider adding support for a more performant, cache-friendly client for collectors in the longer term, and that's potentially something that the [CRI stats migration](https://github.com/kubernetes/enhancements/issues/2371) could benefit from.

**Comment by dims** [comment]
@ehashman can we please open an issue with folks working on it? https://github.com/prometheus/client_golang right?

**Comment by beorn7** [comment]
Yes, please file an issue with the details of this use case.

Back in the days when KSM did their fundamental redesign and optimisation, I thought quite a bit about ways to offer a native way of doing something similar. The efforts got shelved mostly because of two reasons: For one, the KSM use case is rather extreme and rare (and they essentially optimized so thoroughly that they dropped all layers of abstraction, acting on the wire format directly, with all the pros and cons coming with it). The other one is that a major redesign of client_golang has been in the pipeline for quite long, and catering for the "caching" use case would be easier to do after that redesign. But then the redesign still hasn't happened for various reasons…

In any case, having more use cases similar to KSM would definitely inform how such a feature could look like and how useful it would be generally.

**Comment by bwplotka** [comment]
Thanks all. I think book-keeping idea (caching all and garbage collecting when pod is removed) of @smarterclayton @ehashman and others from SIG instrumentation would be the simplest solution here.

I would definitely make sure that the optimization is useful for upcoming CRI-only implementation too, but if I read [the proposal](https://github.com/kubernetes/enhancements/pull/2364) right, the same problem will be there, so we can reuse the same pattern.

Indeed for client_golang, some reusable structure for this "almost" constant series would be nice, but wouldn't it literally look like the garbage collected cached series idea at the end? Do you mind dumping your ideas @beorn7 out at some point, so we can try those once `client_golang` idea v2 is unblocked? (:

**Comment by beorn7** [comment]
A "quick fix" might be possible for the problem at hand, though:

If really most allocations are burned in `prometheus.MakeLabelPairs`, you could indeed avoid it by creating the `prometheus.Desc` once for each label combination encountered and cache it. You have to set the desired labels completely as `constLabels` in `prometheus.NewDesc`, e.g. something like `prometheus.NewDesc("my_metric", "Some help.", nil, prometheus.Labels{"foo": "bar"})`.

Those labels are then pre-rendered in the `Desc` and used whenever you create a `NewConstMetric` with that `Desc`.

(This might be confusing. Happy to provide more help if needed. Just ask…)

**Comment by dgrisonnet** [comment]
From what I can tell, having garbaged collected cached series would cover both the kube-state-metrics and the cadvisor/CRI use cases. This is essentially what we have today in kube-state-metrics, with the exception that we are getting rid of the metrics abstractions and are only storing raw timeseries which add yet another layer of memory optimization.

I think client_golang and the community could benefit to have both data representations since some users might still want to interact with the metrics that are present in the cache, whereas some might just want to regenerate everything without ever needing the layers of abstraction. That would be the case for projects such as kube-state-metrics and cadvisor that are watching particular resources and expose metrics based on their data. They might have more perf/memory gains by regenerating all the timeseries related to a particular object whenever an event is detected rather than storing all the metrics representations in the cache.

Is there already a proposal for the garbage collected cached series idea? Because I'd be interested in participating in the discussions/design. Kube-state-metrics could also be a great guinea pig for this enhancement since we would be able to compare it to the optimization work that was done there.

**Comment by bwplotka** [comment]
Looks like @beorn7 started proposal discussion here: https://github.com/prometheus/client_golang/discussions/917 🤗

**Comment by dims** [comment]
@bwplotka @beorn7 thanks for jumping on this!

**Comment by ehashman** [comment]
Another possible approach I've been working on: offer less cadvisor metrics and split them across some bucketed sub-endpoints to give consumers the choice of which metrics they want (which will hopefully reduce scrape load, to be benchmarked): #109398 

We previously had a proposal to allow cadvisor metric inclusion to be configurable, but I think that would need a KEP: #93650

**Comment by ehashman** [comment]
I've summarized a bunch of possible different ways to optimize CPU usage of cAdvisor in Kubelet in this doc, as well as what work has been done for each: https://docs.google.com/document/d/1TVcAmVmIBxC7BRHaWW5awUPL4jrWKtsiimopOs4zpmw/edit#

**Comment by mrunalp** [comment]
cc: @kolyshkin

**Comment by tossmilestone** [comment]
Any progress on this issue? We also encoutered the same problem.

**Comment by dgrisonnet** [comment]
No, this issue has been stale for a couple of months. There were some concerns regarding the increase in memory usage of kubelet with #108194.

**Comment by LukaszWasko** [comment]
Any progress in this bug? We have the same problem here.

**Comment by SergeyKanzhelev** [comment]
/priority backlog

Seems like this issue is active for a while. Cannot justify it being important-soon
